### DISCUSSION TITLE
Support Coursier credentials out of the box

### DISCUSSION THREAD

**Discussion by laughedelic** [sbt/sbt#6386] [Ideas]
Hi! I've read the docs about setting credentials for sbt:
- https://www.scala-sbt.org/1.x/docs/Publishing.html#Credentials
- https://www.scala-sbt.org/1.x/docs/Proxy-Repositories.html#Using+credentials+for+the+proxy+repository

and they suggest setting up a properties file like `~/.sbt/.cretentials` or `~/.ivy2/.credentials`:

```properties
realm=Sonatype Nexus Repository Manager
host=my.artifact.repo.net
user=admin
password=admin123
```

And then adding a setting like `credentials += Credentials(Path.userHome / ".sbt" / ".credentials")` either in the project or globally.

**Since sbt uses Coursier for dependencies resolution, would it make sense to support Coursier ways of setting credentials and pick them up without any extra setup on the sbt side?**

What I like about Coursier credentials setup is that it expects them in a couple of predefined places, so there's no extra step like with sbt's `credentials += ...` setting.

Coursier supports
- [inline](https://get-coursier.io/docs/other-credentials.html#inline) credentials (e.g. `COURSIER_CREDENTIALS` env var)
- or [property file](https://get-coursier.io/docs/other-credentials.html#property-file) expected at the `~/.config/coursier/credentials.properties` path (or similar on macOS), but in a bit different format:
  ```properties
  simple.username=simple
  simple.password=SiMpLe
  simple.host=artifacts.simple.com
  simple.realm=simple realm
  ```
  (notice `username` vs. `user`, I'm not sure if it's important though)

So basically, if I already have Coursier CLI installed and credentials set in one of these ways, I would like sbt to recognize them out of the box and vice versa: if this was a recommended way to set credentials for sbt and I was later to install Coursier CLI, I would expect no extra setup.

---

I'm posting this here to get some feedback on the general idea. If this is considered desirable, I will be willing to implement it.

**Comment by eed3si9n** [comment]
It sounds like a convenient feature for Coursier CLI users, but I'm also concerned about potential downside as well.
If the existing users are intentionally keeping two different set of credentials or identities in sbt credentials and Coursier credentials, bringing the Coursier credentials implicitly might have unintended consequences. So at least during sbt 1.x, it might be safer to introduce a global setting key like:

```scala
Global / useCoursierCredentials := true
```

or something that people can opt-in?

Also for development and testing it might be useful to have an opt-out as well if it's on by default:

```scala
Global / useCoursierCredentials := false
```

**Reply by laughedelic** [reply]
I understand your concern, but your suggestion to introduce a new setting made me realize that I focused on the wrong thing in my original proposal. I think the core of my idea was not that much about Coursier per se, but about having some default expected locations for credentials and easy ways to set them. Coursier is a convenient reference for this and I thought that since sbt uses Coursier under the hood it would be natural to use the same convention.

My current struggle with sbt is setting credentials in CI (GitHub Actions). Having the credentials, I can't simply put them in a string (like Coursier's inline format: `host(realm) user:password`) and assign to an env var. I need to write a properties file and add an sbt configuration that points to it, either only on CI (which means writing another file) or have a convention in the project to set it to the same thing locally and CI. All this is not difficult, but rather inconvenient to do in a script and has to be repeated for every project's CI setup. 

Let me step back and rephrase my suggestion as two separate things:
1. enable sbt to read credentials in Coursier's format, which is two things:
    - properties file: if I add Coursier's properties file to sbt's `credentials` setting, it would work without the need to change the format of the file
    - [inline](https://get-coursier.io/docs/other-credentials.html#inline): enable sbt to read inline credentials from an env var
2. introduce a convention for where sbt expects its credentials by default: an env var and a file location.

Having just 2. would simplify simple credentials setup because it doesn't require editing sbt settings. And having both 1. and 2. enables the original proposal with a simple env var assignment: `SBT_CREDENTIALS=$COURSIER_CREDENTIALS`, so if I'm using both sbt and Coursier CLI, I can point sbt to Coursier's credentials (explicit opt-in) and expect it to work. If the defaults are different for sbt and Coursier, we can keep distinct sets of credentials for each if it's desired.

What do you think about this approach?

**Reply by jgogstad** [reply]
I think this would be a good improvement. Credentials in a default location is the de-facto way of providing credentials for most modern CLI interfaces (off the top of my head, `git`, `kubectl`, `gcloud`, `aws`, `ssh`).

Cleartext credentials in the file system is still a wart, but having them in a default location makes them easier to work with. I like `git`'s [credential helpers](https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage) which lets me store credentials in for instance the macOS keychain. That's probably a larger endeavour than what you're looking for with this change, and I don't think they are mutually exclusive.

**Reply by eed3si9n** [reply]
@laughedelic Sounds good to me.

I wonder if it would also be useful to support `.netrc` format. That's what Bazel uses - https://www.gnu.org/software/inetutils/manual/html_node/The-_002enetrc-file.html

**Comment by ekrich** [comment]
Having used proxy with basic auth with sbt,  Maven, and Gradle, I think that for the end user setting environment vars in their shell is really is the easiest. In Linux, macOS, and even Windows env vars can be set covering curl and other tools. I find `JAVA_OPTS` with the `-Dprop=val` just works super great and sbt, Metals, coursier all seem to support this. You could make the point that having a password in an env var is a bad idea of course depending on your environment and that a file is slightly better. We even have to deal with passwords that are not fixed but that is handled in the scripts that populate `JAVA_OPTS`.

Adding yet another file and location is not a great idea in my opinion. If anything we should be unifying a location like in `.scala` or something that could be a second place other than the `JAVA_OPTS` location rather than having a file for every tool. This is really what makes the other tools harder than `sbt` is these config files, especially if you have changing properties. Then you also have to write a script to change the properties in all these other locations. This might be an area where working with the scala-cli folks would be a good idea.

We still have a problem with coursier not passing basic auth on redirect which really is a blocker to Scala/sbt adoption in my organization. Changing to ivy resolver works so I know this is the problem.
